### PR TITLE
[berkeley] Add a pickles proof cache 

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR adds a proof cache to pickles, which we can use for tests where we aren't explicitly exercising proof generation.

The cache is designed to
* consume JSON, so that the format is debuggable,
* fallback to generating proofs, if either the format or circuit (according to the verification key) change, and
* make it easy to update the proofs by serialising the final cache back to JSON.

In particular, this approach hooks in directly above the kimchi layer. This means that any circuit logic involved with witness generation is still executed, but the final result is discarded, so we can still use this to check the transaction snark logic is consistent with the 'out-of-circuit' version.

This PR is a no-op, and merely adds the cache infrastructure. This can be used as a baseline to measure improvements in test execution time.

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

* Closes #0000

### CODE REVIEW

**Comment by mrmr1993** [comment]
!ci-build-me

File: src/lib/pickles/wrap.ml [hunk 1]
```diff
@@ -865,10 +865,27 @@ let wrap
...
                 ~auxiliary:auxiliary_inputs pk ~message:next_accumulator
             in
+            let%map.Promise proof =
+              match proof_cache with
+              | None ->
+                  create_proof ()
+              | Some proof_cache -> (
+                  match
+                    Proof_cache.get_wrap_proof proof_cache ~keypair:pk
+                      ~public_input:public_inputs
```
Comment (dannywillems) [inline src/lib/pickles/wrap.ml:879]: Why is indexed only by the public inputs? The same proof will be used even if the witness changes?

Comment (mrmr1993) [reply src/lib/pickles/wrap.ml:879]: Yes. The witness is hidden by the 'zero knowledge' property anyway, so there's no reason not to reuse the proof.

**Review by psteckler — APPROVED** [review]
Transaction snark change looks fine.

**Review by dannywillems — APPROVED** [review]

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Review by nholland94 — APPROVED** [review]

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!approved-for-mainnet

**Pull request merged**